### PR TITLE
Enable most of -Wextra warnings for warnings=all.

### PR DESCRIPTION
**Pull request by marxin**

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -349,7 +349,8 @@ if selected_platform in platform_list:
         if (env["warnings"] == 'extra'):
             env.Append(CCFLAGS=['-Wall', '-Wextra'] + all_plus_warnings + shadow_local_warning)
         elif (env["warnings"] == 'all'):
-            env.Append(CCFLAGS=['-Wall'] + shadow_local_warning)
+            # FIXME: enable -Wimplicit-fallthrough once #26135 is fixed
+            env.Append(CCFLAGS=['-Wall', '-Wextra', '-Wno-implicit-fallthrough', '-Wno-unused-parameter'] + shadow_local_warning)
```
Comment (akien-mga) [inline SConstruct]: It would make more sense to default to `warnings=extra` on GCC/Clang IMO than to add `-Wextra` in the `all` option.

**Comment by akien-mga** [comment]
There seem to be some errors left to fix before we can enable `warnings=extra werror=yes`:
- `-Werror=clobbered` https://travis-ci.org/godotengine/godot/jobs/499164132
- `-Werror=sign-compare` caught by Clang but not GCC

**Comment by marxin** [comment]
Took me couple of iterations to make it green. Can we merge it @akien-mga ?

File: core/typedefs.h [hunk 2]
```diff
@@ -315,4 +315,6 @@ struct _GlobalLock {
 #define _PRINTF_FORMAT_ATTRIBUTE_2_3
 #endif
 
+#define CAST_INT_TO_UCHAR_PTR(ptr) ((uint8_t *)(uintptr_t)(ptr))
```
Comment (akien-mga) [inline core/typedefs.h]: See discussion with @reduz and @hpvb here: http://godot.eska.me/irc-logs/meeting/2019-03-02.log (from 22:20).

Comment (marxin) [reply core/typedefs.h]: I've read it and I must say that first the signature of the function is bad. I would expect an unsigned int as for the offset. Second, I still believe that ```(uint_8 *)0 + some_offsets``` is uglier that usage of ```uintptr_t``` and the casting to ```char *```. Casting to ```char *``` is intentional for aliasing reasons (aliases with a random pointer to e.g. structure). And ```uintptr_t``` is designed to represent a pointer value.

**Comment by marxin** [comment]
May I please remind review of this. I would like to carry on with enablement of additional ones.

File: core/typedefs.h (hunk 2 above)
Comment (marxin) [reply core/typedefs.h]: Good, I've just added a comment ;)

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.1.1.

File: drivers/gles3/rasterizer_canvas_gles3.cpp [hunk 3]
```diff
@@ -2043,7 +2043,7 @@ void RasterizerCanvasGLES3::initialize() {
 		glEnableVertexAttribArray(VS::ARRAY_VERTEX);
 		glVertexAttribPointer(VS::ARRAY_VERTEX, 2, GL_FLOAT, GL_FALSE, sizeof(float) * 4, 0);
 		glEnableVertexAttribArray(VS::ARRAY_TEX_UV);
-		glVertexAttribPointer(VS::ARRAY_TEX_UV, 2, GL_FLOAT, GL_FALSE, sizeof(float) * 4, (float *)0 + 2);
+		glVertexAttribPointer(VS::ARRAY_TEX_UV, 2, GL_FLOAT, GL_FALSE, sizeof(float) * 4, CAST_INT_TO_UCHAR_PTR(2));
```
Comment (akien-mga) [inline drivers/gles3/rasterizer_canvas_gles3.cpp]: This caused #27705. It was a `float *`, not `uint8_t *`.

Comment (akien-mga) [reply drivers/gles3/rasterizer_canvas_gles3.cpp]: #27735 fixes it.

**Pull request merged**